### PR TITLE
Generate: prepare assisted generation for release

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR makes a few final adjustments to assisted generation before its release, including:
1. Merge previously named step 7 [the forward pass after matching with assistant tokens] into step 6 [slicing variables based on the number of matches] -- the variables are already computed in step 3 [selecting the model's next tokens based on the logits], so the code becomes more concise and it helps me explain what's going on more easily. See the (partially bugged) gif below, which is WIP for the blog post.
2. Swaps the order of step 6 [slicing variables] with step 5 [updating the number of candidates for the next iteration] -- makes more sense that the update step for the next iteration is the last one :)
3. Better variable names and improved comments (so the implementation becomes self-documenting)

assisted_gen

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM, thanks! Nice gif!
